mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,1046 @@
+  std::vector<int>
+      loopCarriedDependenceVector; // Loop Carried Dependency vector. A value of
+                                   // "1" at index i represents there is a
+                                   // dependency carried on the loop at depth i
+                                   // in the loopnest
+  llvm::DenseMap<AffineForOp *, long double>
+      numberOfCacheLinesAccessed; // Number of cache lines accessed by each
+                                  // AffineForOp if they were the innermost and
+                                  // only loop in the loopnest
+  std::vector<std::set<Operation *>> referenceGroups;
+};
+} // namespace
+
+/// <summary>
+/// Performs a walk over the loopnest rooted at "forOp" to check if there are
+/// any intermediate "Affine-if" statements in the loopnest.
+/// </summary>
+/// <param name="forOp"></param>
+/// <returns>Returns True if an "Affine-if" statement found.</returns>
+bool hasAffineIfStatement(AffineForOp forOp) {
+  int iffound = 0;
+  forOp.walk([&](Operation *op) {
+    if (isa<AffineIfOp>(op))
+      iffound = 1;
+  });
+  return iffound;
+}
+
+/// <summary>
+/// Tests if a given AffineForOp loopnest has a rectangular shaped iteration
+/// space.
+/// </summary>
+/// <param name="loopnest"></param>
+/// <returns>Returns True if each AffineForOp in the loopnest has a constant
+/// upper and lower bound. Returns false otherwise.</returns>
+bool isRectangularAffineForLoopNest(
+    llvm::SmallVector<AffineForOp, 0> &loopnest) {
+  for (auto a : loopnest) {
+    if (!a.hasConstantUpperBound() || !a.hasConstantLowerBound())
+      return false;
+  }
+  return true;
+}
+
+/// <summary>
+/// Calculates Access Matrix for each memory access operation within the
+/// loopnest rooted at rootForOp by parsing the memory access expressions. For a
+/// memory access Ax+B, it calculates both A and B and populates the protected
+/// "loopAccessMatrices" and "constVector" properties of LoopInterchange object.
+/// </summary>
+/// <param name="rootForOp">Root of the loopnest</param>
+/// <param name="AffineForOpLoopNest"></param>
+/// <param name="loopVarIdMap">The map maintains an Id to each loop induction
+/// variable</param>
+void LoopInterchange::getAffineAccessMatrices(
+    AffineForOp rootForOp,
+    llvm::SmallVector<AffineForOp, 0> &AffineForOpLoopNest,
+    llvm::DenseMap<Value, int64_t> &loopVarIdMap) {
+  // for the mod / ceildiv / floordiv operations, consider them similar to
+  // access to dimensions.
+  this->loopAccessMatrices.clear();
+  this->constVector.clear(); // For Ax+B, this represents B
+  SmallVector<Operation *, 8> loadAndStoreOpInsts;
+  rootForOp.getOperation()->walk([&](Operation *opInst) {
+    if (isa<AffineLoadOp>(opInst) || isa<AffineStoreOp>(opInst)) {
+      loadAndStoreOpInsts.push_back(opInst);
+    }
+  });
+
+  unsigned numOps = loadAndStoreOpInsts.size();
+  for (unsigned i = 0; (i < numOps); ++i) {
+    auto *srcOpInst = loadAndStoreOpInsts[i];
```

Comment:
Drop `inst` suffixes here. `srcOp`. Also avoid using `auto` unless it's obvious and improves readability - it hampers readability here. Spell out the type. Similarly at other places.